File: src/list_articles.js
import { promises as fs } from 'fs';
import { lstatSync as lstat } from 'fs';
import { join } from 'path';

async function list_dir(base_path, dir_path) {
  return (await fs.readdir(join(base_path, dir_path)))
    .map(name => join(dir_path, name));
}

async function find_dir(base_path, dir_path) {
  const items = await list_dir(base_path, dir_path);
  const dirs = items.filter(
    item => lstat(join(base_path, item)).isDirectory());
  return (await Promise.all(dirs.map(dir => find_dir(base_path, dir))))
    .reduce((a, b) => a.concat(b), dirs);
}

async function list_articles(base_path, article_format) {
  const dirs = await find_dir(base_path, '');
  const ext_name = `.${article_format}`;
  const articles_path =
    (await Promise.all(dirs.map(async dir => {
      const items = await list_dir(base_path, dir);
      const files = items.filter(
        item => lstat(join(base_path, item)).isFile());
      return files.filter(filename => filename.endsWith(ext_name));
    }))).reduce((a, b) => a.concat(b), []);
  return {
    dirs, articles_path
  }
}

export default list_articles

File: blog-generator/src/list_articles.rs
use std::path::Path;
use walkdir::WalkDir;

/// Lists all directories and articles in the given path
pub fn list_articles<P: AsRef<Path>>(base_path: P, article_format: &str) -> (Vec<String>, Vec<String>) {
    let base = base_path.as_ref();
    let ext_name = format!(".{}", article_format);
    
    let mut dirs = Vec::new();
    let mut articles_path = Vec::new();
    
    for entry in WalkDir::new(base)
        .min_depth(1)
        .into_iter()
        .filter_map(|e| e.ok())
    {
        let path = entry.path();
        let relative_path = path.strip_prefix(base).unwrap();
        let relative_str = relative_path.to_string_lossy().to_string();
        
        if entry.file_type().is_dir() {
            dirs.push(relative_str);
        } else if entry.file_type().is_file() && relative_str.ends_with(&ext_name) {
            articles_path.push(relative_str);
        }
    }
    
    (dirs, articles_path)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs::{self, File};
    use std::io::Write;
    use tempfile::tempdir;

    #[test]
    fn test_list_articles() {
        let dir = tempdir().unwrap();
        let base = dir.path();
        
        // Create directory structure
        fs::create_dir_all(base.join("subdir")).unwrap();
        File::create(base.join("subdir/test.md")).unwrap()
            .write_all(b"# Test").unwrap();
        
        let (dirs, articles) = list_articles(base, "md");
        
        assert!(dirs.contains(&"subdir".to_string()));
        assert!(articles.contains(&"subdir/test.md".to_string()));
    }
}
